src/Compilers/Core/Portable/Syntax/SyntaxNode.cs

```diff
@@ -479,7 +479,16 @@ public bool ContainsDirective(int rawKind)
                             Debug.Assert(!triviaContainsMatch(current.GetLeadingTriviaCore(), rawKind), "Should not have a match if the token doesn't even have leading trivia");
                         }
 
-                        Debug.Assert(!triviaContainsMatch(current.GetTrailingTriviaCore(), rawKind), "Should never have a match in trailing trivia");
+                        // no need to look within if this token doesn't even have trailing trivia.
+                        if (current.HasTrailingTrivia)
+                        {
+                            if (triviaContainsMatch(current.GetTrailingTriviaCore(), rawKind))
```

Comment:
> In the problematic scenario, `#if` is in the middle of the line. Perhaps the parser is not supposed to treat it as a directive?

Sound good, I haven't realized that's an invariant. Thanks.

Reply:
> I haven't realized that's an invariant. 

I am not stating this is a fact, but the nature of the assert makes me think that is probably what is going on.

Reply:
Yes.

> If that is correct, it might be reasonable to assume that a directive can never be in a trailing trivia. In the problematic scenario, #if is in the middle of the line. Perhaps the parser is not supposed to treat it as a directive?

I strongly align with this approach. 

I believe we intended to have the invariant that directives were only leading trivia.  As such, if they do appear at the end, while it is fine to be *resilient* to seeing it (And give good messages) we shoudl still represent the problematic code in some other way (like simply making it into 'skipped token trivia').

Reply:
It looks like we have this:

```c#
                    case '#':
                        if (_allowPreprocessorDirectives)
                        {
                            this.LexDirectiveAndExcludedTrivia(afterFirstToken, isTrailing || !onlyWhitespaceOnLine, ref triviaList);
                            break;
                        }
                        else
                        {
                            return;
                        }
```

We can either say that we don't go into LexDirectiveAndExcludedTrivia if isTrailing is false.  Or we go in, get the result, but then convert somehow into a skipped 'bad token' that acts as the trailing trivia of hte current token.  I'm fine with either approach.

Reply:
I've tried this approach but it seems some parts of Roslyn assume there are directives in trailing trivia, for example this test stops working:

https://github.com/dotnet/roslyn/blob/09a7b388974b48af9ef28a539de10cb463014e15/src/Workspaces/CSharpTest/Formatting/FormattingTests.cs#L5757-L5773

I'm not sure if we should continue with this approach and try to make it work, or we could say that directives can be in trailing trivia in error scenarios - and make `SyntaxNode.ContainsDirective` API ignore those (so we would just remove the assert).

Reply:
It's fine with me for that test to stop working.